services/datamanager/builtin/sync/upload_data_capture_file.go

```diff
@@ -125,55 +180,161 @@ func getImagesTimestamps(res *pb.GetImagesResponse, sensorData *v1.SensorData) (
+	md *datasyncPB.UploadMetadata,
+	sd *datasyncPB.SensorData,
+) error {
+	// if the binary sensor data has a mime type, set the file extension
+	// to match
+	fileExtensionFromMimeType := getFileExtFromMimeType(sd.GetMetadata().GetMimeType())
+	if fileExtensionFromMimeType != "" {
+		md.FileExtension = fileExtensionFromMimeType
+	}
+	if _, err := client.DataCaptureUpload(ctx, &datasyncPB.DataCaptureUploadRequest{
+		Metadata:       md,
+		SensorContents: []*datasyncPB.SensorData{sd},
+	}); err != nil {
+		return errors.Wrap(err, "DataCaptureUpload failed")
+	}
 
-		toUpload := sensorData[0]
+	return nil
+}
 
-		// First send metadata.
-		streamMD := &v1.StreamingDataCaptureUploadRequest_Metadata{
-			Metadata: &v1.DataCaptureUploadMetadata{
-				UploadMetadata: uploadMD,
-				SensorMetadata: toUpload.GetMetadata(),
-			},
-		}
-		if err := c.Send(&v1.StreamingDataCaptureUploadRequest{UploadPacket: streamMD}); err != nil {
-			return errors.Wrap(err, "StreamingDataCaptureUpload failed sending metadata")
+func uploadMultipleBinarySensorData(
+	ctx context.Context,
+	client datasyncPB.DataSyncServiceClient,
+	uploadMD *datasyncPB.UploadMetadata,
+	sensorData []*datasyncPB.SensorData,
+	path string,
+	logger logging.Logger,
+) error {
+	// this is the common case
+	if len(sensorData) == 1 {
+		logger.Debugf("attempting to upload small binary file using DataCaptureUpload, sensor data, file: %s", path)
+		return uploadBinarySensorData(ctx, client, uploadMD, sensorData[0])
+	}
+
+	// we only go down this path if the capture method returned multiple binary
+	// responses, which at time of writing, only includes camera.GetImages data.
+	for i, sd := range sensorData {
+		// Otherwise use the unary endpoint
+		logger.Debugf("attempting to upload small binary file using DataCaptureUpload, sensor data index: %d, ext: %s, file: %s", i, path)
+		// we clone as the uploadMD may be changed for each sensor data
+		// and I'm not confident that it is safe to reuse grpc request structs
+		// between calls if the data in the request struct changes
+		clonedMD := proto.Clone(uploadMD).(*datasyncPB.UploadMetadata)
+		if err := uploadBinarySensorData(ctx, client, clonedMD, sd); err != nil {
+			return err
 		}
+	}
+	return nil
+}
+
+func uploadMultipleLargeBinarySensorData(
+	ctx context.Context,
+	client datasyncPB.DataSyncServiceClient,
+	uploadMD *datasyncPB.UploadMetadata,
+	sensorData []*datasyncPB.SensorData,
+	path string,
+	logger logging.Logger,
+) error {
+	if len(sensorData) == 1 {
+		logger.Debugf("attempting to upload large binary file using StreamingDataCaptureUpload, sensor data file: %s", path)
+		return uploadLargeBinarySensorData(ctx, client, uploadMD, sensorData[0], path, logger)
+	}
 
-		// Then call the function to send the rest.
-		if err := sendStreamingDCRequests(ctx, c, toUpload.GetBinary(), path, logger); err != nil {
-			return errors.Wrap(err, "StreamingDataCaptureUpload failed to sync")
+	for i, sd := range sensorData {
```

Comment:
This is a good point. This problem is that we want to check if our data sync request is above or below the limit, which, to your point, isn't the same thing as the capture file size. 

It is also not the same thing as the size of the sensorData turned into proto (as that doesn't contain the upload metadata).

Yes I would prefer to use streaming endpoints for all uploads, which would require an API change as we don't have that endpoint today for tabular data.

It is also a good idea to change to a different wire format for tabular data (maybe bison) as otherwise, you can't predict the size of the proto data without serializing it twice.

I do think this is best left for a later change.